osutil/epoll/epoll_test.go

```diff
@@ -0,0 +1,466 @@
+package epoll_test
+
+import (
+	"os"
+	"syscall"
+	"testing"
+	"time"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/snapcore/snapd/osutil/epoll"
+
+	. "gopkg.in/check.v1"
+)
+
+func Test(t *testing.T) { TestingT(t) }
+
+type epollSuite struct{}
+
+var _ = Suite(&epollSuite{})
+
+func (*epollSuite) TestString(c *C) {
+	c.Check(epoll.Readable.String(), Equals, "Readable")
+	c.Check(epoll.Writable.String(), Equals, "Writable")
+	c.Check(epoll.Readiness(epoll.Readable|epoll.Writable).String(), Equals, "Readable|Writable")
+}
+
+func (*epollSuite) TestOpenClose(c *C) {
+	e, err := epoll.Open()
+	c.Assert(err, IsNil)
+	c.Assert(e.Fd() == -1, Equals, false)
+	c.Assert(e.Fd() == 0, Equals, false)
+	c.Assert(e.Fd() == 1, Equals, false)
+	c.Assert(e.Fd() == 2, Equals, false)
+	c.Assert(e.RegisteredFdCount(), Equals, 0)
+
+	err = e.Close()
+	c.Assert(err, IsNil)
+	c.Assert(e.Fd(), Equals, -1)
+}
+
+func concurrentlyRegister(e *epoll.Epoll, fd int, errCh chan error) {
+	err := e.Register(fd, epoll.Readable)
+	errCh <- err
+}
+
+func concurrentlyDeregister(e *epoll.Epoll, fd int, errCh chan error) {
+	err := e.Deregister(fd)
+	errCh <- err
+}
+
+func waitMillisecondsThenWriteToFd(msec int, fd int, msg []byte) error {
+	time.Sleep(time.Duration(msec) * time.Millisecond)
+	_, err := unix.Write(fd, msg)
+	return err
+}
+
+func (*epollSuite) TestRegisterWaitModifyDeregister(c *C) {
+	e, err := epoll.Open()
+	c.Assert(err, IsNil)
+
+	socketFds, err := unix.Socketpair(unix.AF_UNIX, unix.SOCK_STREAM, 0)
+	c.Assert(err, IsNil)
+
+	listenerFd := socketFds[0]
+	senderFd := socketFds[1]
+
+	err = unix.SetNonblock(listenerFd, true)
+	c.Assert(err, IsNil)
+
+	err = e.Register(listenerFd, epoll.Readable)
+	c.Assert(err, IsNil)
+
+	msg := []byte("foo")
+
+	go waitMillisecondsThenWriteToFd(1000, senderFd, msg)
+
+	events, err := e.Wait()
+	c.Assert(err, IsNil)
+	c.Assert(len(events), Equals, 1)
+	c.Assert(events[0].Fd, Equals, listenerFd)
+
+	buf := make([]byte, len(msg))
+	_, err = unix.Read(events[0].Fd, buf)
+	c.Assert(err, IsNil)
+	c.Assert(buf, DeepEquals, msg)
+
+	err = e.Modify(listenerFd, epoll.Readable|epoll.Writable)
+	c.Assert(err, IsNil)
+
+	err = e.Deregister(listenerFd)
+	c.Assert(err, IsNil)
+
+	err = e.Close()
+	c.Assert(err, IsNil)
+}
+
+func (*epollSuite) TestRegisterUnhappy(c *C) {
+	e, err := epoll.Open()
+	c.Assert(err, IsNil)
+
+	// attempt to register a standard file
+	newFile, err := os.CreateTemp("/tmp", "snapd-TestRegisterUnhappy-")
+	c.Assert(err, IsNil)
+	defer newFile.Close()
+	defer os.Remove(newFile.Name())
+	err = e.Register(int(newFile.Fd()), epoll.Readable)
+	c.Check(err, Equals, syscall.Errno(0x1)) // "operation not permitted"
+
+	// attempt to register nonexistent FD
+	err = e.Register(42, epoll.Readable)
+	c.Check(err, Equals, syscall.Errno(0x9)) // "bad file descriptor"
```

Comment:
this is failing for me and also in the CI run

Reply:
Interesting, it worked for me in the standalone `go test -check.f ...` call, and I think it succeeded with a full `./run-checks` run, but you're right, it's failing for `./run-checks` for me now as well. I think the fact that I create a bunch of sockets as part of the other tests means the (arbitrary) FD 42 is actually a socket that is able to be successfully registered. I'll bump it to a larger number which should be beyond what we would expect to seel

Reply:
I see, but also see my new comment about closing those test sockets as well

Reply:
Sounds good. I addressed the abandoned sockets problem, and I'm going to leave this a large number in case other unit tests cause many FDs to be opened. Thanks!